### PR TITLE
Fix dual and add lattice equality

### PR DESCRIPTION
**Pull request by thofma**
For lattice equality I am leaving a lot on the table performance wise, so please complain if it is too slow :)

### CODE REVIEW

**Comment by StevellM** [comment]
For equality, we could also create another function, of this kind (I did on my own repository, I was trying to implement equality of AbsLat): 

``` 
@doc Markdown.doc"""
    pseudo_hnf(L::AbsLat) -> AbsLat

Given a lattice $L$ with pseudo-matrix $M$, return the pseudo-hermite normal 
form of $M$, and cache it as the pseudo-matrix of $L$.
"""
function pseudo_hnf(L::AbsLat)
  M = L.pmat
  new_mat = pseudo_hnf(M, :lowerleft)
  L.pmat = new_mat
  return new_mat
end
```
so once computed for `L`, we cache the pseudo-hermite normal form of its `pseudo_matrix` and so one could add a line in the test of equality like

```
if L.pmat == M.pmat
  return true
end
```
so that if their respective pseudo-matrix are already in pseudo hnf form, we avoid having to recompute them (for instance if we run a code where we need to compare several lattices to a same fixed one)

**Comment by thofma** [comment]
Thanks! Unfortunately we do not have uniqueness for pseudo-Hermite normal forms, but the caching will still help. I will amend the PR.

**Comment by StevellM** [comment]
Yes I remarked this, I could take the `pseudo_hnf` of the `pseudo_hnf` of a pseudo matrix and get another result. I guess your `_modules_equality` is here for this non-uniqueness ?

**Comment by thofma** [comment]
Yes (it also assumes `pseudo_hnf(..., :lowerleft)` :)

**Comment by simonbrandhorst** [comment]
@StevellM  note that caching is okay. But mutating an object is dangerous and in general discouraged. For instance some properties of the lattice may depend on the pseudomatrix, for instance the automorphism group generators can be given with respect to the given pseudo basismatrix. If the pseudo basis silently changes, that invalidates prior computations and leads to bugs.

**Comment by thofma** [comment]
Yes, anybody can access fields of objects and change them. I am not sure how this relates to caching. Touching fields of objects is done at your own risk.

**Comment by StevellM** [comment]
Yes that's true, I always forget not to touch to those access... How should we proceed then ? create a new access for this pseudo matrix in pseudo hnf, or make sure that the pseudo matrix cached at the creation of the lattice is already in pseudo hnf ?

**Comment by thofma** [comment]
I will fix it tomorrow by storing it explicitly.

**Review by simonbrandhorst — APPROVED** [review]

**Comment by simonbrandhorst** [comment]
> I will fix it tomorrow by storing it explicitly.
I suggest the interface  `pseudo_hnf(L::AbsLat)`

**Comment by simonbrandhorst** [comment]
@StevellM `dual(L)` should require that `L` is non-degenerate (otherwise the dual is not a lattice). Please add an `@req` check.

File: src/QuadForm/Herm/Lattices.jl [hunk 1]
```diff
@@ -321,14 +320,15 @@ end
 ################################################################################
 
 function dual(L::HermLat)
-  G, B = _gram_schmidt(gram_matrix_of_rational_span(L), involution(L))
+  G = gram_matrix_of_rational_span(L)
+  @req rank(G) == nrows(G) "Lattice must be non-degenerate"
```
Comment (simonbrandhorst) [inline src/QuadForm/Herm/Lattices.jl:324]:
```
 rank(L::AbsLat) -> Int

  Returns the rank of the underlying module of L.
```
Hence this is not the correct test. I propose introducing a (cached) method `isdegenerate(L::AbsLat)` which returns `rank(L) == rank(rational_span(L))`

Comment (thofma) [reply src/QuadForm/Herm/Lattices.jl:324]: I am testing whether the Gram matrix has full rank. Why is this not correct?

**Comment by thofma** [comment]
Ah, annoying. Can we use `issubset`? There the ordering is the other way around. I know this is an annoying inconsistency.

File: src/QuadForm/Herm/Lattices.jl (hunk 1 above)
Comment (simonbrandhorst) [reply src/QuadForm/Herm/Lattices.jl:324]: ah I missred the lhs for `rank(L)` so nvm it is correct

**Comment by simonbrandhorst** [comment]
We can enable issubset as well.

**Comment by thofma** [comment]
I will fix it later today.

File: test/QuadForm/Lattices.jl [hunk 2]
```diff
@@ -110,14 +110,39 @@
   o = @inferred Hecke.automorphism_group_order(L)
   @test o == 1536
 
+  @test ambient_space(P*L) === ambient_space(L)
+
+  @test issublattice(L, P*L)
+  @test !issublattice(P * L, L)
+  @test issubset(P*L, L)
+  @test !issubset(L, P*L)
+  @test isintegral(L) == issublattice(dual(L), :)
```
Comment (thofma) [inline test/QuadForm/Lattices.jl:119]:
```suggestion
  @test isintegral(L) == issublattice(dual(L), L)
```

**Pull request merged**